### ISSUE TITLE
Bug: Middleware http pipeline is constructed wrong

### ISSUE THREAD

**Issue by jjcollinge** [dapr/dapr#948]
/area runtime
> /area operator
> /area placement
> /area docs
> /area test-and-release

## What version of Dapr?

master

## Expected Behavior
When registering middleware as a component it should be included in the HTTP pipeline once
i.e. `Cors` --> `CustomMiddleware` --> `Router`

## Actual Behavior
When registering custom middleware as a component it is included in the HTTP pipeline twice.
i.e. `CustomMiddleware` --> `Cors` --> `CustomMiddleware` --> `Router`

## Steps to Reproduce the Problem
Register the oAuth middleware component, configure dapr to use it and then put a break point in the handler. Make a request to the `1.HelloWorld sample app` via the Dapr endpoint and you'll see it invoked 4 times. twice for the client request and twice for the app request.

## Diagnosis
This occurs due to the way middleware is being applied in the HTTP server [here](https://github.com/dapr/dapr/blob/d17e9243b308e830649b0bf3af5f6e84fd543baf/pkg/http/server.go#L50)

```
1. handler := s.pipeline.Apply(router.HandleRequest)
2. go func() {
3.	if s.tracingSpec.Enabled {
4.		log.Fatal(fasthttp.ListenAndServe(fmt.Sprintf(":%v", s.config.Port),
                                s.pipeline.Apply(diag.TracingHTTPMiddleware(s.tracingSpec, corsHandler.CorsMiddleware(handler)))))
5.	} else {
6.		log.Fatal(fasthttp.ListenAndServe(fmt.Sprintf(":%v", s.config.Port),
				s.pipeline.Apply(corsHandler.CorsMiddleware(handler))))
7.	}
8. }()
```

Line 1 passes the `router.HandleRequest` into the `pipeline.Apply(...)` which will build up a pipeline like so:
`handler` = `CustomMiddleware.2` --> `CustomMiddleware.1` --> `router.HandleRequest`

Let's assume `s.tracingSpec.Enabled = false` so line 6 executes next. It will pass the existing `handler` into the `corsHandler.CorsMiddleware` effectively making the pipeline:
`handler` = `corsHandler.CorsMiddleware` --> `CustomMiddleware.2` --> `CustomMiddleware.1` --> `router.HandleRequest`
__Note:__ This is the pipeline we want and so should stop here.

However, it then passes the resulting handler into `pipeline.Apply(...)`  again, which then iterates the custom middleware handlers again, finally giving you a pipeline as:
`handler` = `CustomMiddleware.2` --> `CustomMiddleware.1` -->  `corsHandler.CorsMiddleware` --> `CustomMiddleware.2` --> `CustomMiddleware.1` --> `router.HandleRequest`

**Comment by jjcollinge** [comment]
I've implemented a bit of a refactor as per above in the PR #951 rather than in this fix.

**Comment by yaron2** [comment]
Closed with #949
